README.md
# cloom

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2020 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2020-05-02
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2020-05-02
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/cloom/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/cloom/compare/0.1.0...0.1.1


test/cloom/core_test.clj
(ns cloom.core-test
  (:require [clojure.test :refer :all]
            [cloom.core :refer :all]))

(deftest decify-test
  (testing "decify fail."
    (is (= (decify "00") 0))
    (is (= (decify "10") 16))
    (is (= (decify "ff") 255))
    ))

(deftest hexify-test
  (testing "hexify fail."
    (is (= (hexify 0) "00"))
    (is (= (hexify 16) "10"))
    (is (= (hexify 255) "ff"))
    ))

(deftest to-int32-test
  (testing "to-int32 fail."
    (is (= (to-int32 1) (list 1 0 0 0)))
    (is (= (to-int32 256) (list 0 1 0 0)))
    (is (= (to-int32 65536) (list 0 0 1 0)))
    (is (= (to-int32 65537) (list 1 0 1 0)))
    (is (= (to-int32 16777216) (list 0 0 0 1)))
    (is (= (to-int32 16777217) (list 1 0 0 1)))
    ))

(deftest from-int32-test
  (testing "from-int32 fail."
    (is (= (from-int32 (list 1 0 0 0)) 1))
    (is (= (from-int32 (list 0 1 0 0)) 256))
    (is (= (from-int32 (list 0 0 1 0)) 65536))
    (is (= (from-int32 (list 1 0 1 0)) 65537))
    (is (= (from-int32 (list 0 0 0 1)) 16777216))
    (is (= (from-int32 (list 1 0 0 1)) 16777217))
    ))

(deftest from-int16-test
  (testing "from-int16 fail."
    (is (= (from-int16 (list 1 0)) 1))
    (is (= (from-int16 (list 0 1)) 256))
    (is (= (from-int16 (list 255 255)) 65535))
    ))

(deftest to-int16-test
  (testing "to-int16 fail."
    (is (= (to-int16 1) (list 1 0)))
    (is (= (to-int16 256) (list 0 1)))
    (is (= (to-int16 65535) (list 255 255)))
    ))

(deftest hexstring-to-ints-test
  (testing "hexstring-to-ints fail."
    (is (= (hexstring-to-ints "0f1022") (list 15 16 34)))
    (is (= (hexstring-to-ints "49574144670b0000b822de000000001f") (list 73 87 65 68 103 11 0 0 184 34 222 0 0 0 0 31)))
    ))

(deftest read-header-test
  (testing "read-header fail."
    (is (= (read-header "49574144670b0000b822de000000001f")
           {:id "IWAD"
            :num-lumps 2919
            :info-table-offset 14557880}
           ))
    ))

(deftest read-directory-test
  (testing "read-directory fail."
    (is (= (count (read-directory (slurp "the_assault.wad.hex") (read-header (slurp "the_assault.wad.hex"))))
           51))
    ))

(deftest things-to-ints-test
  (testing "things-to-ints fail."
    (let [thing {:x-position 1234
                 :y-position 2345
                 :angle-facing 1
                 :thing-type 2035
                 :flags 7}
          ]
      (is (= (count (things-to-ints (list thing))) bytes-per-thing))
      (is (= thing (read-thing (things-to-ints (list thing)))))
      )))

(deftest linedefs-to-ints-test
  (testing "linedefs-to-ints fail."
    (let [linedef {:start-vertex 0
                   :end-vertex 1
                   :flags 4
                   :special-type 1
                   :sector-tag 4
                   :front-sidedef 1
                   :back-sidedef 2
                   }
          ]
      (is (= (count (linedefs-to-ints (list linedef))) bytes-per-linedef))
      (is (= linedef (read-linedef (linedefs-to-ints (list linedef)))))
      )))

(deftest sidedefs-to-ints-test
  (testing "sidedefs-to-ints fail."
    (let [sidedef {:x-offset 0
                   :y-offset 1
                   :upper-texture-name "TEXTURE1"
                   :lower-texture-name "TEXTURE2"
                   :middle-texture-name "TEXTURE3"
                   :sector-number 2
                   }
          ]
      (is (= (count (sidedefs-to-ints (list sidedef))) bytes-per-sidedef))
      (is (= sidedef (read-sidedef (sidedefs-to-ints (list sidedef)))))
      )))

(deftest vertexes-to-ints-test
  (testing "vertexes-to-ints fail."
    (let [vertex {:x-position 0
                  :y-position 1
                  }
          ]
      (is (= (count (vertexes-to-ints (list vertex))) bytes-per-vertex))
      (is (= vertex (read-vertex (vertexes-to-ints (list vertex)))))
      )))

(deftest sectors-to-ints-test
  (testing "sectors-to-ints fail."
    (let [sector {:floor-height 0
                  :ceiling-height 1
                  :floor-texture-name "TEXTURE1"
                  :ceiling-texture-name "TEXTURE2"
                  :light-level 3
                  :sector-type 5
                  :tag-number 5
                  }
          ]
      (is (= (count (sectors-to-ints (list sector))) bytes-per-sector))
      (is (= sector (read-sector (sectors-to-ints (list sector)))))
      )))

(deftest map-to-wad-hexstring-test
  (testing "map-to-wad-hexstring fail."
    (let [hs (slurp "the_assault.wad.hex")
          header (read-header hs)
          directory (read-directory hs header)
          lumps (map #(read-lump hs %) directory)
          things (nth lumps 3)
          linedefs (nth lumps 4)
          sidedefs (nth lumps 5)
          vertexes (nth lumps 6)
          sectors (nth lumps 10)
          out-hs (map-to-wad-hexstring "MAP01" things linedefs sidedefs vertexes sectors)
          ]
      (spit "test_copied.wad.hex" out-hs))
    ))

(deftest custom-map-test
  (testing "Custom map test fail."
    (let [things (list 
                   ; player start
                   {:x-position 16
                    :y-position 16
                    :angle-facing 0
                    :thing-type 1 
                    :flags 0x7
                    }
                   ; barrel
                   {:x-position 64
                    :y-position 64
                    :angle-facing 0
                    :thing-type 2035
                    :flags 0x7
                    })
          linedefs (list
                     ; SW-NW
                     {:start-vertex 0
                      :end-vertex 1
                      :flags 0x1
                      :special-type 0
                      :sector-tag 0
                      :front-sidedef 0
                      :back-sidedef 0xffff ;none
                      }
                     ; NW-NE
                     {:start-vertex 1
                      :end-vertex 2
                      :flags 0x1
                      :special-type 0
                      :sector-tag 0
                      :front-sidedef 1
                      :back-sidedef 0xffff ;none
                      }
                     ; NE-SE
                     {:start-vertex 2
                      :end-vertex 3
                      :flags 0x1
                      :special-type 0
                      :sector-tag 0
                      :front-sidedef 2
                      :back-sidedef 0xffff ;none
                      }
                     ; NE-SW
                     {:start-vertex 3
                      :end-vertex 0
                      :flags 0x1
                      :special-type 0
                      :sector-tag 0
                      :front-sidedef 3
                      :back-sidedef 0xffff ;none
                      })
          sidedefs (list
                     ; SW-NW
                     {:x-offset 0
                      :y-offset 0
                      :upper-texture-name ""
                      :lower-texture-name ""
                      :middle-texture-name "BRONZE1"
                      :sector-number 0
                      }
                     ; NW-NE
                     {:x-offset 0
                      :y-offset 0
                      :upper-texture-name ""
                      :lower-texture-name ""
                      :middle-texture-name "BRONZE2"
                      :sector-number 0
                      }
                     ; NE-SE
                     {:x-offset 0
                      :y-offset 0
                      :upper-texture-name ""
                      :lower-texture-name ""
                      :middle-texture-name "TEKGREN1"
                      :sector-number 0
                      }
                     ; SE-SW
                     {:x-offset 0
                      :y-offset 0
                      :upper-texture-name ""
                      :lower-texture-name ""
                      :middle-texture-name "TEKGREN2"
                      :sector-number 0
                      })
          vertexes (list
                     ; SW
                     {:x-position 0
                      :y-position 0}
                     ; NW
                     {:x-position 0 
                      :y-position 256}
                     ; NE
                     {:x-position 256
                      :y-position 256}
                     ; SE
                     {:x-position 256
                      :y-position 0}
                     )
          sectors (list
                    {:floor-height 0
                     :ceiling-height 64
                     :floor-texture-name "FLAT2"
                     :ceiling-texture-name "CEIL1_2"
                     :light-level 255
                     :sector-type 0
                     :tag-number 0
                     })
          hs (map-to-wad-hexstring "MAP01" things linedefs sidedefs vertexes sectors)
          ]
      (is (= nil (println (str "Thing-ints:\n" (into [] (things-to-ints things))))))
      (is (= nil (println (str "Linedef-ints:\n" (into [] (linedefs-to-ints linedefs))))))
      (is (= nil (println (str "Sidedef-ints:\n" (into [] (sidedefs-to-ints sidedefs))))))
      (is (= nil (println (str "Vertex-ints:\n" (into [] (vertexes-to-ints vertexes))))))
      (is (= nil (println (str "Sector-ints:\n" (into [] (sectors-to-ints sectors))))))
      (is (not= nil hs))
      (is (= nil (spit "test_room.wad.hex" hs)))
    )))



src/cloom/core.clj
(ns cloom.core)

(def bytes-per-thing 10)
(def bytes-per-linedef 14)
(def bytes-per-sidedef 30)
(def bytes-per-vertex 4)
(def bytes-per-sector 26)

(defn hexify [x]
  "Convert a 0-255 number to hexadecimal, msb first."
  (let [letters [\0 \1 \2 \3 \4 \5 \6 \7 \8 \9 \a \b \c \d \e \f]
        msb (nth letters (int (/ x 16)))
        lsb (nth letters (rem x 16))
        ]
    (str msb lsb)))

(defn decify [x]
  "Convert a two character list of hexadecimals (msb first) to a single int."
  (let [letters [\0 \1 \2 \3 \4 \5 \6 \7 \8 \9 \a \b \c \d \e \f]
        values (apply hash-map (interleave letters (range 16)))
        msb (values (first x))
        lsb (values (nth x 1))
        ]
    (+ (* msb 16) lsb)))

(defn to-int32 [x]
  "Return as four bytes, lsb first."
  (map #(rem (bit-shift-right x (* % 8)) 256) (range 4)))

(defn from-int32 [bs]
  "Take list of four ints (lsb first) and turn into a single int."
  (reduce + (map #(bit-shift-left (nth bs %) (* 8 %)) (range 4))))

(defn from-int16 [bs]
  "Take list of four ints (lsb first) and turn into a single int."
  (reduce + (map #(bit-shift-left (nth bs %) (* 8 %)) (range 2))))

(defn to-int16 [x]
  "Return as four bytes, lsb first."
  (map #(rem (bit-shift-right x (* % 8)) 256) (range 2)))

(defn clean-hexstring [hs]
  "Return hexstring with all spaces, newlines, and [^0-9a-f] removed."
  (filter #(#{\0 \1 \2 \3 \4 \5 \6 \7 \8 \9 \a \b \c \d \e \f} %) hs))

(defn string-to-padded-ints [s]
 "Convert a string to 8 ints zero-padded."
 (concat (map int s)
         (for [i (range (- 8 (count s)))] 0)))

(defn hexstring-to-ints [hexstring]
  "Read a hexdump (generated by xxd -p) and return a list of ints."
  (let [
        hs (clean-hexstring hexstring)
        hexbytes (map #(take 2 (drop % hs)) (range 0 (count hs) 2))
        ]
    (map decify hexbytes)))

(defn ints-to-hexstring [byte-ints]
  "Convert a list of ints into a hexadecimal string."
  (apply str (map hexify byte-ints)))

(defn read-header [hexstring]
  "Reads the header from the wad as a list of ints."
  (let [hs (clean-hexstring (take 32 hexstring))
        wad-ints (hexstring-to-ints hs)
        id (apply str (map char (take 4 wad-ints)))
        num-lumps (from-int32 (take 4 (drop 4 wad-ints)))
        info-table-offset (from-int32 (take 4 (drop 8 wad-ints)))
        ;dummy (println (str "id: " id))
        ]
    {:id id
     :num-lumps num-lumps
     :info-table-offset info-table-offset
     }))

(defn read-directory [hexstring header]
  "Reads the directory, given the offset and the wad-ints."
  (let [offset (* (:info-table-offset header) 2)
        num-lumps (:num-lumps header)
        hs (clean-hexstring hexstring)
        directory-hexstring (drop offset hs)
        directory-ints (hexstring-to-ints directory-hexstring)
        ]
    (for [lump (range num-lumps)]
      (let [i (* lump 16)
            lump-offset (from-int32 (take 4 (drop i directory-ints)))
            lump-size (from-int32 (take 4 (drop (+ 4 i) directory-ints)))
            lump-name (apply str (map char (take 8 (drop (+ 8 i) directory-ints))))
            ]
        {:offset lump-offset
         :size lump-size
         :name lump-name}))))

(defn read-thing [thing-ints]
  "Reads a 10 byte (in int format) thing."
  (let [x-position (from-int16 (take 2 thing-ints))
        y-position (from-int16 (take 2 (drop 2 thing-ints)))
        angle-facing (from-int16 (take 2 (drop 4 thing-ints)))
        thing-type (from-int16 (take 2 (drop 6 thing-ints)))
        flags (from-int16 (take 2 (drop 8 thing-ints)))
        ]
    {:x-position x-position
     :y-position y-position
     :angle-facing angle-facing
     :thing-type thing-type
     :flags flags
     }))

(defn read-linedef [linedef-ints]
  "Reads a 14 byte (in int format) linedef."
  (let [start-vertex (from-int16 (take 2 linedef-ints))
        end-vertex (from-int16 (take 2 (drop 2 linedef-ints)))
        flags (from-int16 (take 2 (drop 4 linedef-ints)))
        special-type (from-int16 (take 2 (drop 6 linedef-ints)))
        sector-tag (from-int16 (take 2 (drop 8 linedef-ints)))
        front-sidedef (from-int16 (take 2 (drop 10 linedef-ints)))
        back-sidedef (from-int16 (take 2 (drop 12 linedef-ints)))
        ]
    {:start-vertex start-vertex
     :end-vertex end-vertex
     :flags flags
     :special-type special-type
     :sector-tag sector-tag
     :front-sidedef front-sidedef
     :back-sidedef back-sidedef
     }))

(defn read-sidedef [sidedef-ints]
  "Reads a 30 byte (in int format) sidedef."
  (let [x-offset (from-int16 (take 2 sidedef-ints))
        y-offset (from-int16 (take 2 (drop 2 sidedef-ints)))
        upper-texture-name (re-find #"^[A-Za-z0-9-]*" (apply str (map char (take 8 (drop 4 sidedef-ints)))))
        lower-texture-name (re-find #"^[A-Za-z0-9-]*" (apply str (map char (take 8 (drop 12 sidedef-ints)))))
        middle-texture-name (re-find #"^[A-Za-z0-9-]*" (apply str (map char (take 8 (drop 20 sidedef-ints)))))
        sector-number (from-int16 (take 2 (drop 28 sidedef-ints)))
        ]
    {:x-offset x-offset
     :y-offset y-offset
     :upper-texture-name upper-texture-name
     :lower-texture-name lower-texture-name
     :middle-texture-name middle-texture-name
     :sector-number sector-number
     }))

(defn read-vertex [vertex-ints]
  "Reads a 4 byte (in int format) vertex."
  (let [x-position (from-int16 (take 2 vertex-ints))
        y-position (from-int16 (take 2 (drop 2 vertex-ints)))
        ]
    {:x-position x-position
     :y-position y-position
     }))

(defn read-sector [sector-ints]
  "Reads a 26 byte (in int format) sector."
  (let [floor-height (from-int16 (take 2 sector-ints))
        ceiling-height (from-int16 (take 2 (drop 2 sector-ints)))
        floor-texture-name (re-find #"^[A-Za-z0-9-]*" (apply str (map char (take 8 (drop 4 sector-ints)))))
        ceiling-texture-name (re-find #"^[A-Za-z0-9-]*" (apply str (map char (take 8 (drop 12 sector-ints)))))
        light-level (from-int16 (take 2 (drop 20 sector-ints)))
        sector-type (from-int16 (take 2 (drop 22 sector-ints)))
        tag-number (from-int16 (take 2 (drop 24 sector-ints)))
        ]
    {:floor-height floor-height
     :ceiling-height ceiling-height
     :floor-texture-name floor-texture-name
     :ceiling-texture-name ceiling-texture-name
     :light-level light-level
     :sector-type sector-type
     :tag-number tag-number
     }))

(defn parse-entries
  "Read a lump with a given reader function, whose entries are entry-size long."
  ([reader entry-size lump-ints] (parse-entries reader entry-size lump-ints (list )))
  ([reader entry-size lump-ints entries]
   (if (empty? lump-ints)
     entries
     (recur reader
            entry-size
            (drop entry-size lump-ints)
            (concat entries (list (reader lump-ints)))))))


(defn read-lump [hexstring directory-entry]
  "Reads the hexstring for a lump given a directory entry."
  (let [hs (take (* 2 (:size directory-entry))
                 (drop (* 2 (:offset directory-entry)) (clean-hexstring hexstring)))
        lump-ints (hexstring-to-ints hs)
        lump-name (re-find #"^[A-Z0-9]*" (:name directory-entry))
        ]
    (case lump-name
      "THINGS" (parse-entries read-thing bytes-per-thing lump-ints)
      "LINEDEFS" (parse-entries read-linedef bytes-per-linedef lump-ints)
      "SIDEDEFS" (parse-entries read-sidedef bytes-per-sidedef lump-ints)
      "VERTEXES" (parse-entries read-vertex bytes-per-vertex lump-ints)
      "SEGS" "SEGS not implemented."
      "SSECTORS" "SSECTORS not implemented."
      "NODES" "NODES not implemented."
      "SECTORS" (parse-entries read-sector bytes-per-sector lump-ints)
      "REJECT" "REJECT not implemented."
      "BLOCKMAP" "BLOCKMAP not implemented."
      "BEHAVIOR" "BEHAVIOR not implemented."
      (str "Unrecognized lump type: " lump-name)
      )
    ))

; write functions

(defn things-to-ints [things]
  "Take a list of things and turn them into a list of ints representing each byte in file."
  (let [f (fn [thing]
            (concat
              (to-int16 (:x-position thing))
              (to-int16 (:y-position thing))
              (to-int16 (:angle-facing thing))
              (to-int16 (:thing-type thing))
              (to-int16 (:flags thing))))
        ]
    (apply concat (map f things))))

(defn linedefs-to-ints [linedefs]
  "Take a list of linedefs and turn them into a list of ints representing each byte in file."
  (let [f (fn [linedef]
            (concat
              (to-int16 (:start-vertex linedef))
              (to-int16 (:end-vertex linedef))
              (to-int16 (:flags linedef))
              (to-int16 (:special-type linedef))
              (to-int16 (:sector-tag linedef))
              (to-int16 (:front-sidedef linedef))
              (to-int16 (:back-sidedef linedef))))
        ]
    (apply concat (map f linedefs))))

(defn sidedefs-to-ints [sidedefs]
  "Take a list of sidedefs and turn them into a list of ints representing each byte in file."
  (let [f (fn [sidedef]
            (concat
              (to-int16 (:x-offset sidedef))
              (to-int16 (:y-offset sidedef))
              (string-to-padded-ints (:upper-texture-name sidedef))
              (string-to-padded-ints (:lower-texture-name sidedef))
              (string-to-padded-ints (:middle-texture-name sidedef))
              (to-int16 (:sector-number sidedef))))
        ]
    (apply concat (map f sidedefs))))

(defn vertexes-to-ints [vertexes]
  "Take a list of vertexes and turn them into a list of ints representing each byte in file."
  (let [f (fn [vertex]
            (concat
              (to-int16 (:x-position vertex))
              (to-int16 (:y-position vertex))))
        ]
    (apply concat (map f vertexes))))

(defn sectors-to-ints [sectors]
  "Take a list of sectors and turn them into a list of ints representing each byte in file."
  (let [f (fn [sector]
            (concat
              (to-int16 (:floor-height sector))
              (to-int16 (:ceiling-height sector))
              (string-to-padded-ints (:floor-texture-name sector))
              (string-to-padded-ints (:ceiling-texture-name sector))
              (to-int16 (:light-level sector))
              (to-int16 (:sector-type sector))
              (to-int16 (:tag-number sector))))
        ]
    (apply concat (map f sectors))))

(defn map-to-wad-hexstring
  "Convert a map to hexstring of a wad file."
  [map-name things linedefs sidedefs vertexes sectors]
  (let [thing-ints (things-to-ints things)
        linedef-ints (linedefs-to-ints linedefs)
        sidedef-ints (sidedefs-to-ints sidedefs)
        vertex-ints (vertexes-to-ints vertexes)
        sector-ints (sectors-to-ints sectors)
        header-size 12
        num-lumps 6 ; includes nodebuilt lumps (6), above lumps (5), map name (1)
        map-start header-size
        things-start header-size
        linedefs-start (+ things-start (count thing-ints))
        sidedefs-start (+ linedefs-start (count linedef-ints))
        vertexes-start (+ sidedefs-start (count sidedef-ints))
        sectors-start (+ vertexes-start (count vertex-ints))
        directory-start (+ sectors-start (count sector-ints))
        directory-ints (concat (to-int32 map-start) (to-int32 0) (string-to-padded-ints "MAP01")
                               (to-int32 things-start) (to-int32 (count thing-ints)) (string-to-padded-ints "THINGS")
                               (to-int32 linedefs-start) (to-int32 (count linedef-ints)) (string-to-padded-ints "LINEDEFS")
                               (to-int32 sidedefs-start) (to-int32 (count sidedef-ints)) (string-to-padded-ints "SIDEDEFS")
                               (to-int32 vertexes-start) (to-int32 (count vertex-ints)) (string-to-padded-ints "VERTEXES")
                               (to-int32 sectors-start) (to-int32 (count sector-ints)) (string-to-padded-ints "SECTORS")
                               )
        header-ints (concat (map int "PWAD")
                            (to-int32 num-lumps)
                            (to-int32 directory-start))
        dummy (println (str "map-start:  " map-start
                            "\nthings-start:  " things-start
                            "\nlinedefs-start:  " linedefs-start
                            "\nsidedefs-start:  " sidedefs-start
                            "\nvertexes-start:  " vertexes-start
                            "\nsectors-start:  " sectors-start
                            "\ndirectory-start:  " directory-start
                            "\ndirectory-ints:\n" directory-ints))
        ]
    (ints-to-hexstring (concat header-ints thing-ints linedef-ints sidedef-ints vertex-ints sector-ints directory-ints))
    ))




src/cloom/turing.clj
(ns cloom.turing
  (:require [cloom.core :refer :all]))

(defn logic-node []
  "Build a custom map that demonstrates a single logic node."
  ; walk from the NW corner to the SE across a square room composed of two sectors, W and E
  ; if the E sector is raised, you'll slide to the south wall, passing triggers along the way
  ; if the E sector is not raised, you'll reach the SE corner, passing triggers along the way
  ; at the end of either path, teleport to next logic node (identical destination)
  (let [things (list
                 ; player start
                 {:x-position 16
                  :y-position 496
                  :angle-facing 315 ; we have to start at an angle, because you can only "slide" of NS/EW walls for some reason
                  :thing-type 1 
                  :flags 0x7
                  }
                 ; teleport destination (same as player start)
                 {:x-position 16
                  :y-position 496
                  :angle-facing 315 ; we have to start at an angle, because you can only "slide" of NS/EW walls for some reason
                  :thing-type 14
                  :flags 0x7
                  }
                 )
        linedefs (list
                   ; SW-NW wall
                   {:start-vertex 0
                    :end-vertex 1
                    :flags 0x1
                    :special-type 211 ; can switch platform height, repeatable toggle
                    :sector-tag 1
                    :front-sidedef 0
                    :back-sidedef 0xffff ;none
                    }
                   ; NW-N wall
                   {:start-vertex 1
                    :end-vertex 2
                    :flags 0x1
                    :special-type 0
                    :sector-tag 0
                    :front-sidedef 1
                    :back-sidedef 0xffff ;none
                    }
                   ; N-S divide
                   {:start-vertex 2
                    :end-vertex 3
                    :flags 0x4 ; two-sided
                    :special-type 0
                    :sector-tag 0
                    :front-sidedef 2
                    :back-sidedef 4 ; two-sided
                    }
                   ; S-SW wall
                   {:start-vertex 3
                    :end-vertex 0
                    :flags 0x1
                    :special-type 0
                    :sector-tag 0
                    :front-sidedef 3
                    :back-sidedef 0xffff ;none
                    }
                   ; N-NE wall
                   {:start-vertex 2
                    :end-vertex 4
                    :flags 0x1
                    :special-type 0
                    :sector-tag 0
                    :front-sidedef 5
                    :back-sidedef 0xffff ;none
                    }
                   ; NE-SE wall
                   {:start-vertex 4
                    :end-vertex 5
                    :flags 0x1
                    :special-type 0
                    :sector-tag 0
                    :front-sidedef 6
                    :back-sidedef 0xffff ;none
                    }
                   ; SE-S wall
                   {:start-vertex 5
                    :end-vertex 3
                    :flags 0x1
                    :special-type 0
                    :sector-tag 0
                    :front-sidedef 7 
                    :back-sidedef 0xffff ;none
                    }
                   ; teleport region (faces outward)
                   {:start-vertex 0
                    :end-vertex 5
                    :flags 0x0
                    :special-type 97; teleporter
                    :sector-tag 0
                    :front-sidedef 8
                    :back-sidedef 12
                    }
                   ; teleport region (faces outward)
                   {:start-vertex 5
                    :end-vertex 7
                    :flags 0x0
                    :special-type 97; teleporter
                    :sector-tag 0
                    :front-sidedef 9
                    :back-sidedef 13
                    }
                   ; W teleport region (faces outward)
                   {:start-vertex 7
                    :end-vertex 6
                    :flags 0x0
                    :special-type 97; teleporter
                    :sector-tag 0
                    :front-sidedef 10
                    :back-sidedef 14
                    }
                   ; W teleport region (faces outward)
                   {:start-vertex 6
                    :end-vertex 0
                    :flags 0x0
                    :special-type 97; teleporter
                    :sector-tag 0
                    :front-sidedef 11
                    :back-sidedef 15
                    }
                   )
        sidedefs (list
                   ; room SW-NW (0)
                   {:x-offset 0
                    :y-offset 0
                    :upper-texture-name "-"
                    :lower-texture-name "-"
                    :middle-texture-name "BRONZE1"
                    :sector-number 0
                    }
                   ; NW-N
                   {:x-offset 0
                    :y-offset 0
                    :upper-texture-name "-"
                    :lower-texture-name "-"
                    :middle-texture-name "BRONZE1"
                    :sector-number 0
                    }
                   ; N-S
                   {:x-offset 0
                    :y-offset 0
                    :upper-texture-name "BRONZE2"
                    :lower-texture-name "BRONZE1"
                    :middle-texture-name "-"
                   :sector-number 0
                    }
                   ; S-SW
                   {:x-offset 0
                    :y-offset 0
                    :upper-texture-name "-"
                    :lower-texture-name "-"
                    :middle-texture-name "TEKGREN2"
                    :sector-number 0
                    }
                   ; S-N
                   {:x-offset 0
                    :y-offset 0
                    :upper-texture-name "BRONZE2"
                    :lower-texture-name "BRONZE1"
                    :middle-texture-name "-"
                    :sector-number 1
                    }
                   ; N-NE
                   {:x-offset 0
                    :y-offset 0
                    :upper-texture-name "ASHWALL3"
                    :lower-texture-name "CEMENT9"
                    :middle-texture-name "BRONZE2"
                    :sector-number 1
                    }
                   ; NE-SE
                   {:x-offset 0
                    :y-offset 0
                    :upper-texture-name "ASHWALL3"
                    :lower-texture-name "CEMENT9"
                    :middle-texture-name "TEKGREN1"
                    :sector-number 1
                    }
                   ; SE-S
                   {:x-offset 0
                    :y-offset 0
                    :upper-texture-name "ASHWALL3"
                    :lower-texture-name "CEMENT9"
                    :middle-texture-name "BRONZE2"
                    :sector-number 1
                    }
                   ; teleport region w outward (8)
                   {:x-offset 0
                    :y-offset 0
                    :upper-texture-name "TEKGREN2"
                    :lower-texture-name "CEMENT9"
                    :middle-texture-name "-"
                    :sector-number 2
                    }
                   ; teleport region s outward
                   {:x-offset 0
                    :y-offset 0
                    :upper-texture-name "TEKGREN2"
                    :lower-texture-name "CEMENT9"
                    :middle-texture-name "-"
                    :sector-number 2
                    }
                   ; teleport region e outward
                   {:x-offset 0
                    :y-offset 0
                    :upper-texture-name "TEKGREN2"
                    :lower-texture-name "CEMENT9"
                    :middle-texture-name "-"
                    :sector-number 2
                    }
                   ; teleport region n outward
                   {:x-offset 0
                    :y-offset 0
                    :upper-texture-name "TEKGREN2"
                    :lower-texture-name "CEMENT9"
                    :middle-texture-name "-"
                    :sector-number 2
                    }
                   ; teleport region w inward (12)
                   {:x-offset 0
                    :y-offset 0
                    :upper-texture-name "TEKGREN2"
                    :lower-texture-name "CEMENT9"
                    :middle-texture-name "-"
                    :sector-number 2
                    }
                   ; teleport region s inward
                   {:x-offset 0
                    :y-offset 0
                    :upper-texture-name "TEKGREN2"
                    :lower-texture-name "CEMENT9"
                    :middle-texture-name "-"
                    :sector-number 2
                    }
                   ; teleport region e inward
                   {:x-offset 0
                    :y-offset 0
                    :upper-texture-name "TEKGREN2"
                    :lower-texture-name "CEMENT9"
                    :middle-texture-name "-"
                    :sector-number 2
                    }
                   ; teleport region n inward
                   {:x-offset 0
                    :y-offset 0
                    :upper-texture-name "TEKGREN2"
                    :lower-texture-name "CEMENT9"
                    :middle-texture-name "-"
                    :sector-number 2
                    }
                   )
        vertexes (list
                   ; SW of room (0)
                   {:x-position 0
                    :y-position 0}
                   ; NW of room (1)
                   {:x-position 0
                    :y-position 512}
                   ; N of room (2)
                   {:x-position 256
                    :y-position 512}
                   ; S of room (3)
                   {:x-position 256
                    :y-position 0}
                   ; NE of room (4)
                   {:x-position 512
                    :y-position 512}
                   ; SE of room (5)
                   {:x-position 512
                    :y-position 0}
                   ; NW teleport region (6)
                   {:x-position 0
                    :y-position 32}
                   ; NE teleport region (7)
                   {:x-position 512
                    :y-position 32}
                   )
        sectors (list
                  ; W
                  {:floor-height 0
                   :ceiling-height 256
                   :floor-texture-name "FLOOR0_1"
                   :ceiling-texture-name "CEIL1_2"
                   :light-level 255
                   :sector-type 0
                   :tag-number 0
                   }
                  ; E
                  {:floor-height 0 ; can be toggled to ceiling
                   :ceiling-height 256
                   :floor-texture-name "FLAT5"
                   :ceiling-texture-name "CEIL1_3"
                   :light-level 255
                   :sector-type 0
                   :tag-number 1
                   }
                  ; teleporter pad
                  {:floor-height 16
                   :ceiling-height 256
                   :floor-texture-name "GATE1"
                   :ceiling-texture-name "CEIL1_2"
                   :light-level 255
                   :sector-type 0
                   :tag-number 0
                   }
                  )
        hs (map-to-wad-hexstring "MAP01" things linedefs sidedefs vertexes sectors)
        ]
    (spit "test_room.wad.hex" hs)
  ))

